sql/driver/driver_test.go

```diff
@@ -227,3 +227,60 @@ func TestInsecure(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestTransactions(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	s, db := setup(t)
+	defer cleanup(s, db)
+
+	if _, err := db.Exec(`CREATE DATABASE t`); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec(`CREATE TABLE t.kv (k CHAR PRIMARY KEY, v CHAR)`); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec(`INSERT INTO t.kv VALUES ($1, $2), ($3, $4)`, "a", "b", "c", nil); err != nil {
+		t.Fatal(err)
+	}
+
+	if rows, err := db.Query("SELECT * FROM t.kv"); err != nil {
+		t.Fatal(err)
+	} else {
+		results := readAll(t, rows)
+		expectedResults := asResultSlice([][]string{
+			{"k", "v"},
+			{"a", "b"},
+			{"c", ""},
+		})
+		expectedResults[2][1] = nil
+		if err := verifyResults(expectedResults, results); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	txn, err := db.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := txn.Exec(`DELETE FROM t.kv WHERE k IN ($1)`, "c"); err != nil {
+		t.Fatal(err)
+	}
+
+	if rows, err := txn.Query("SELECT * FROM t.kv"); err != nil {
+		t.Fatal(err)
+	} else {
+		results := readAll(t, rows)
+		expectedResults := asResultSlice([][]string{
+			{"k", "v"},
+			{"a", "b"},
+		})
+		if err := verifyResults(expectedResults, results); err != nil {
+			t.Fatal(err)
+		}
+	}
+	// TODO(vivek): Merge with https://github.com/cockroachdb/cockroach/pull/2045
+	// and remove isError().
+	if err := txn.Commit(); !isError(err, "cannot commit a read-only transaction") {
```

Comment:
Ok. You're still going to have this problem though unless you do all of the work in the same round-trip. If I type the following shell session
```sql
BEGIN TRANSACTION; // sent to server
UPDATE ... // sent
COMMIT TRANSACTION; // sent
```

then when the `COMMIT TRANSACTION` is executed, there needs to be the information that this transaction actually wrote data in order to determine whether to send a kv-level `EndTransaction`.

It's not going to be an issue if we don't allow "hanging" transactions.